src/table/boot.rs

```diff
@@ -112,22 +110,27 @@ pub struct BootServices {
     calculate_crc32: usize,
 
     // Misc services
-    copy_mem: extern "win64" fn(dest: *mut u8, src: *const u8, len: usize),
-    set_mem: extern "win64" fn(buffer: *mut u8, len: usize, value: u8),
+    copy_mem: unsafe extern "win64" fn(dest: *mut u8, src: *const u8, len: usize),
+    set_mem: unsafe extern "win64" fn(buffer: *mut u8, len: usize, value: u8),
 
     // New event functions (UEFI 2.0 or newer)
     create_event_ex: usize,
 }
 
 impl BootServices {
     /// Raises a task's priority level and returns its previous level.
-    pub fn raise_tpl(&self, tpl: Tpl) -> Tpl {
-        (self.raise_tpl)(tpl)
-    }
-
-    /// Restores a task’s priority level to its previous value.
-    pub fn restore_tpl(&self, old_tpl: Tpl) {
-        (self.restore_tpl)(old_tpl)
+    ///
+    /// The effect of calling raise_tpl with a Tpl that is below the current one
+    /// (which, sadly, cannot be queried) is undefined by the UEFI spec, which
+    /// also warns against remaining at high Tpls for extended periods of time.
+    ///
+    /// This function outputs an RAII guard that will automatically restore the
+    /// original Tpl when dropped.
+    pub unsafe fn raise_tpl(&self, tpl: Tpl) -> TplGuard<'_> {
+        TplGuard {
+            boot_services: self,
+            old_tpl: (self.raise_tpl)(tpl),
+        }
     }
```

Comment:
As it turns out, `raise_tpl` has a surprisingly unsafe contract, which noticeably calls for every `raise_tpl` to be associated with a matching `restore_tpl`. An RAII guard seems appropriate for this.